onnxruntime/core/optimizer/identity_elimination.cc

```diff
@@ -10,16 +10,81 @@
 
 namespace onnxruntime {
 
+/**
+  Special case to eliminate Identity node when its output is graph output 
+  note that there is no output edge for Identity
+
+  X ---> Identity ---> graph output
+
+  becomes 
+
+  X ---> graph output
+ */
 Status EliminateIdentity::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
-  if (graph_utils::RemoveNode(graph, node)) {
-    rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
-  }
+  // regular case - when Identity's output has no graph output 
+  if (graph.GetNodeOutputsInGraphOutputs(node).empty()) {
+    if (graph_utils::RemoveNode(graph, node)) {
+      rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+    }
+  } else {
+    // handling special case
 
+    // keep a reference of output def to the graph output
+    NodeArg* output = node.MutableOutputDefs()[0];
+    const Node* p_input_node = graph_utils::GetInputNode(node, 0);
+    // get mutable input node
+    Node& input_node = *graph.GetNode(p_input_node->Index());
+    int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(input_node, node.MutableInputDefs()[0]->Name());
+    // remove Identity node and its input edge
+    graph.RemoveNode(node.Index());
+    // update input node's output def to the graph output
+    input_node.MutableOutputDefs()[output_idx] = output;
+    rule_effect = RewriteRuleEffect::kRemovedCurrentNode; 
+  }
   return Status::OK();
 }
 
 bool EliminateIdentity::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger& logger) const {
-  return graph_utils::CanRemoveNode(graph, node, logger);
+  if (graph_utils::CanRemoveNode(graph, node, logger)) {
+    return true;
+  }
+  
+  // check the input nodearg of Identity - it must only have one consumer, which is the Identity itself
+
+  const Node* p_input_node = graph_utils::GetInputNode(node, 0);
+  if (p_input_node == nullptr)
+    return false;
+    
+  // find the edge between input node and this Identity node, and then get its src arg from input node
```

Comment:
you can use graph_utils::GetNodeOutputIndexFromOutputName

Reply:
GetNodeOutputIndexFromOutputName() won't find the output name from output def for the Identity on the right. We need to rely on src arg index in the edges/relations to figure this out. 

![image](https://user-images.githubusercontent.com/4484531/114947515-26064380-9e02-11eb-9f57-612de65f1a7d.png)

Reply:
will double check ...

Reply:
adding it back, it was not working due to the side effect of GetConsumerNodes()